### PR TITLE
Deployscript

### PR DESCRIPTION
**Pull request by jdotpz**
* Adds a deploy-socorro.sh main script in ./bin
* Adds a few dependent scripts under deploylib (see README.md in ./bin/ for more info
* Updates AMI of buildbox

### CODE REVIEW

**Comment by jdotpz** [comment]
Output of a full test run: https://leeroy.mocotoolsstaging.net/job/STAGE-socorromaster/29/consoleFull
## Approximate timing
* Build/rpm creation: *5 minutes*
* AMI creation: *12 minutes*
* Terraform applies: *30 seconds*
* Node swap to new AMI: *5 minutes*

File: bin/README.md [hunk 1]
```diff
@@ -0,0 +1,105 @@
+# Deployment scripts for socorro
+
+The scripts here used to deploy socorro applications are:
+
+* ./deploy-socorro.sh  *Main script, passed an argument of stage or prod.  Calls a number of scripts in the deploylib subdirectory*
```
Comment (phrawzty) [inline bin/README.md]: Style nit: paths and filenames should be `fixed width` which then allows the description to *not* be italicised (italics should just be for *emphasis*).

Same for below.

File: bin/README.md [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+* ./deploylib/create_ami.sh  *Uses packer scripts from ../packer to generate an AMI, tag it with the socorro git has*
+
+* ./deploylib/create_rpm.sh  *Creates an updated RPM package with fpm, signs it, and uploads it to the S3 public repo*
+
+* ./deploylib/identify_app.sh  *A script which, given an appname-envname, will identify the ELB, AS Group, and terraform name to use*
+
+* ./deploylib/stage-socorro-master.list  *A text list of all the appname-envnames a socorro deploy must touch*
+
+
+## Setup for a builder box/jenkins box
```
Comment (phrawzty) [inline bin/README.md]: Consistent naming is important; to wit, it's buildbox (not builder box).

File: bin/README.md [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+
+A few steps are required to prepare Jenkins or another CI server to properly use this script.
+
+### The AWS cli must be installed
+
+*curl "https://s3.amazonaws.com/aws-cli/awscli-bundle.zip" -o "awscli-bundle.zip"*
```
Comment (phrawzty) [inline bin/README.md]:
```
Style nit: CLI recipes should be in code blocks.
```

File: bin/README.md [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+
+*unzip awscli-bundle.zip*
+
+*sudo ./awscli-bundle/install -i /usr/local/aws -b /usr/local/bin/aws*
+
+
+### An IAM user for jenkins must be created and granted certain permissions
+
```
Comment (phrawzty) [inline bin/README.md]: `s/Rights to//g;`

File: bin/README.md [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+* Rights to Read/Write to the public S3 rpm repo (for syncing the updated socorro rpms)
+
+* Rights to AWS EC2 API's (* granted)
+
+* Rights to AWS IAM PassRole (to apply roles to newly created infrastructure)
+
+
+### A configuration file for AWS creds created
+
```
Comment (phrawzty) [inline bin/README.md]: Here again, style points for `filenames` and
```
config files
```

File: bin/README.md [hunk 6]
```diff
@@ -0,0 +1,105 @@
...
+### Add PATH locations in /home/centos/.bash_profile
+
+* Added /home/centos/terraform to path in .bash_profile
+
+* Added /usr/pgsql-9.3/bin/ to path in .bash_profile
+
+* Added /usr/local/bin/ to path in .bash_profile
+
+
+### Jenkins installed on the builderbox
```
Comment (phrawzty) [inline bin/README.md]: Buildbox

File: bin/README.md [hunk 7]
```diff
@@ -0,0 +1,105 @@
...
+
+
+### Add PATH locations in /home/centos/.bash_profile
+
+* Added /home/centos/terraform to path in .bash_profile
+
+* Added /usr/pgsql-9.3/bin/ to path in .bash_profile
+
+* Added /usr/local/bin/ to path in .bash_profile
+
```
Comment (phrawzty) [inline bin/README.md]: Not sure where "Added" comes from here; maybe just list the paths and state that they need to be on the `$PATH`? Let the user decide how they want to handle that.

File: bin/README.md [hunk 8]
```diff
@@ -0,0 +1,105 @@
...
+
+### Install and configure terraform/packer
+
+* Installed into /home/centos/terraform
+
+* Added all terraform variable symbolic links
+
+* Installed packer
+
```
Comment (phrawzty) [inline bin/README.md]: Not sure what the lines above refer to. Seems like personal notes, perhaps? We'll need to either flesh this out or replace with a *TODO* placeholder.

Same for the two sections below.

File: bin/deploy-socorro.sh [hunk 9]
```diff
@@ -0,0 +1,233 @@
...
+############################
+# COMMON FUNCTIONS AND SETTINGS
+# Bail on any fail
+set -e
+
+# Get incoming arguments
+REQUIREDARGS=1
+NUMOFARGS=$#
+ENVNAME=$1
+RANDOM_STRING=`date +%s%m%d%SS|sed 's/ //g'`
```
Comment (phrawzty) [inline bin/deploy-socorro.sh]: Bash has a built in `$RANDOM` [function](http://www.tldp.org/LDP/abs/html/randomvar.html) which will generate a 5-digit long random number - this is the canonical way to generate random strings in Bash.
```bash
RANDOM_STRING="$RANDOM-$RANDOM"
```

File: bin/deploy-socorro.sh [hunk 10]
```diff
@@ -0,0 +1,233 @@
...
+# Get AWS creds injected
+. /home/centos/.aws-config
+
+# Bring in functions in deploylib
+. /home/centos/socorro-infra/bin/deploylib/identify_app.sh
+. /home/centos/socorro-infra/bin/deploylib/create_rpm.sh
+. /home/centos/socorro-infra/bin/deploylib/create_ami.sh
+
+# Show usage instructions for noobs
+show_usage() {
```
Comment (phrawzty) [inline bin/deploy-socorro.sh]: Please prefix functions with `function`.
```bash
function show_usage() {
```

File: bin/deploy-socorro.sh [hunk 11]
```diff
@@ -0,0 +1,233 @@
...
+    echo " =================== ";echo " ";
+    echo " Syntax "
+    echo " ./deploy-socorro.sh $env (stage or prod)"
+    echo " "
+    echo " Example: ./deploy-socorro.sh stage"
+    exit 0
+    }
```
Comment (phrawzty) [inline bin/deploy-socorro.sh]: Please align the closing parens with the opening line.
```bash
function show_usage() {
    return 0
}
```

Same for all functions below.

File: bin/deploy-socorro.sh [hunk 12]
```diff
@@ -0,0 +1,233 @@
...
+
+# Check if you are asking for halp or passing the wrong number of args
+check_syntax() {
+  if [ "$1" = "-h" ] || [ "$1" = "--help" ]
```
Comment (phrawzty) [inline bin/deploy-socorro.sh]: Please use 4 space indenting throughout.

Please place the `then` on the same line.
```bash
function funkytown() {
    if [ condition ]; then
        return 0
    fi
}

File: bin/deploy-socorro.sh [hunk 13]
```diff
@@ -0,0 +1,233 @@
...
+    then
+    show_usage
+  fi
+  if [ ${NUMOFARGS} -ne ${REQUIREDARGS} ]
+    then
+    echo "`date` -- ERROR, ${REQUIREDARGS} arguments required and ${NUMOFARGS} is an invalid number of args"
```
Comment (phrawzty) [inline bin/deploy-socorro.sh]: :+1: for datestamp!

File: bin/README.md [hunk 14]
```diff
@@ -0,0 +1,101 @@
...
+* Rights to AWS IAM PassRole (to apply roles to newly created infrastructure)
+
+
+### A configuration file for AWS creds created
+
+```/home/centos/.aws-config```
+
+* Perms set to 755, containing the following:
+
+```export AWS_ACCESS_KEY_ID=KEY```
```
Comment (phrawzty) [inline bin/README.md]: These should be in a single block.

**Comment by phrawzty** [comment]
You use the "application" nomenclature throughout; however, as noted previously, our established nomenclature for this is "role", which was chosen because it encompasses a greater field of meaning than "app".  I realise that this comes across as a nit, but it's important to be consistent with these things - the half-drunk engineer[0] who's trying to fix burning infra at 4 in the morning will thank us for not confusing them by naming things in multiple ways. :wine_glass: :blush: 

[0] read: *future-us*.

File: terraform/jenkins_wrapper.sh [hunk 15]
```diff
@@ -0,0 +1,102 @@
+#!/usr/bin/env bash
+# This script acts as a wrapper to Terraform.
```
Comment (phrawzty) [inline terraform/jenkins_wrapper.sh]: This appears to be a copy of `wrapper.sh` with a different name. Is this on purpose?

**Comment by phrawzty** [comment]
And, of course, this needs an *epic* squashing. :game_die:

**Comment by jdotpz** [comment]
Squashed!

**Comment by rhelmer** [comment]
r+

File: bin/README.md (hunk 8 above)
Comment (phrawzty) [inline bin/README.md]: @rhelmer There isn't an official RPM, no; however, it'd be trivial to make one ([bug 1161969](https://bugzilla.mozilla.org/show_bug.cgi?id=1161969)).

**Comment by phrawzty** [comment]
# :tada:

**Pull request merged**